flang/lib/Lower/Bridge.cpp

```diff
@@ -1324,69 +1328,148 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+  }
+
+  Fortran::lower::ExValue genExprEleAddr(const Fortran::lower::SomeExpr &expr,
+                                         llvm::ArrayRef<mlir::Value> lcvs) {
+    return createSomeExtendedAddress(toLocation(), *this, expr, localSymbols,
+                                     lcvs);
+  }
+
   /// Shared for both assignments and pointer assignments.
-  void genFIR(const Fortran::evaluate::Assignment &assignment) {
+  void genFIR(const Fortran::evaluate::Assignment &assign) {
     std::visit(
         Fortran::common::visitors{
             [&](const Fortran::evaluate::Assignment::Intrinsic &) {
+              auto loc = toLocation();
               const auto *sym =
-                  Fortran::evaluate::UnwrapWholeSymbolDataRef(assignment.lhs);
-              if (sym && Fortran::semantics::IsAllocatable(*sym)) {
-                // Assignment of allocatable are more complex, the lhs
-                // may need to be deallocated/reallocated.
-                // See Fortran 2018 10.2.1.3 p3
-                TODO();
-              } else if (sym && Fortran::semantics::IsPointer(*sym)) {
-                // Target of the pointer must be assigned.
-                // See Fortran 2018 10.2.1.3 p2
-                auto lhsType = assignment.lhs.GetType();
-                assert(lhsType && "lhs cannot be typeless");
-                if (isNumericScalarCategory(lhsType->category())) {
-                  auto val = genExprValue(assignment.rhs);
-                  auto addr = genExprValue(assignment.lhs);
-                  auto toTy = fir::dyn_cast_ptrEleTy(addr.getType());
-                  auto cast =
-                      builder->convertWithSemantics(toLocation(), toTy, val);
-                  builder->create<fir::StoreOp>(toLocation(), cast, addr);
-                } else if (isCharacterCategory(lhsType->category())) {
-                  TODO();
-                } else {
-                  assert(lhsType->category() ==
-                         Fortran::common::TypeCategory::Derived);
-                  TODO();
-                }
-              } else if (assignment.lhs.Rank() > 0) {
+                  Fortran::evaluate::UnwrapWholeSymbolDataRef(assign.lhs);
+              // Assignment of allocatable are more complex, the lhs may need to
+              // be deallocated/reallocated. See Fortran 2018 10.2.1.3 p3
+              const bool isHeap =
+                  sym && Fortran::semantics::IsAllocatable(*sym);
+              // Target of the pointer must be assigned. See Fortran
+              // 2018 10.2.1.3 p2
+              const bool isPointer = sym && Fortran::semantics::IsPointer(*sym);
+              auto lhsType = assign.lhs.GetType();
+              assert(lhsType && "lhs cannot be typeless");
+
+              if (assign.lhs.Rank() > 0 || (assign.rhs.Rank() > 0 && isHeap)) {
                 // Array assignment
                 // See Fortran 2018 10.2.1.3 p5, p6, and p7
+                auto shape = getShape(assign.lhs);
```

Comment:
The compiler implementations I looked at use the shape of the LHS entity being stored into.

GetShape() will return expressions (not just constants).

The bridge does not lower Fortran interfaces at all. This PR does not implement mini-contexts. That work is TODO.

Reply:
Right, GetShape() returns expressions, but in the case of `x + foo()` with foo returning an relocatable, I am not sure what expression `GetShape()` can return, at best, it would return an expression that reads that calls `foo()` which we have to be careful not to execute twice.
All that to say, `GetSahpe()` makes sens in expression analysis/semantics, but I am not sure it can succeed in all cases where lowering needs a shape.
I agree `GetSahpe()` will work with simple array expressions and may be OK to start with. Maybe a small note could be added that it is not a silver bullet and cannot be used for expressions that need to be evaluated to compute the shape.